fontforge/featurefile.c

```diff
@@ -456,20 +456,21 @@ static OTLookup *lookup_in_rule(struct fpst_rule *r,int seq,int *index, int *pos
     /* That's legal in otf, but I don't think it can be specified in the */
     /*  feature file. It doesn't seem likely to be used so I ignore it */
 
-    for ( i=0; i<r->lookup_cnt && seq<r->lookups[i].seq; ++i );
-    if ( i>=r->lookup_cnt )
-return( NULL );
+    for ( i=0; i<r->lookup_cnt && seq>r->lookups[i].seq; ++i );
+    if ( i > 0 && seq != r->lookups[i].seq )
+        --i;
     *index = i;
     *pos = seq-r->lookups[i].seq;
     otl = r->lookups[i].lookup;
     if ( seq==r->lookups[i].seq )
-return( otl );
-    if ( otl->lookup_type == gsub_ligature )
-return( otl );
+        return( otl );
+    /* This does not seem to be needed, and it is better if not used at all */
+    //if ( otl->lookup_type == gsub_ligature )
```

Comment:
It causes `sub a f'lookup lig1 t' i';` to be written as `sub a f'lookup lig1 t'lookup lig1 i'lookup lig1;`, assuming lookup lig1 has a rule `sub f t i by f_t_i`

Reply:
@skef
- The function `lookup_in_rule()` should return the lookup that affects the sequence position passed as a parameter `seq`.
- The line `for ( i=0; i<r->lookup_cnt && seq<r->lookups[i].seq; ++i );` is looping over all lookups in the rule `r` until we find the lookup we want (which has a sequence position that equals the passed `seq`).
- So we need to increment `i` as long as the lookups we are looping over have sequences less than the passed `seq` or in other words keep looping over lookups as long as `seq` is greater than `lookup.seq`.
- For a sanity check, let's assume we didn't make that change and we had a rule that had 2 contextual lookups: one affecting sequence position 0 and the other affeting seq pos 1, when we pass `seq==1` to the function (looking for the second lookup) `i` will never get incremented because `1<r->lookups[0].seq` which is `1<0` is false at the first iteration of the loop, so we will never get the second lookup which we want.
- After making the change, `1>r->lookups[0].seq` which is `1>0` will evaluate to true and we will increment `i` until we find a lookup that has a `lookup.seq` that is equal or greater than the passed `seq`, then we decrement `i` once if `lookup.seq != seq` to get the desired lookup.

Reply:
Ok, that more or less makes sense to me.

How did you determine the ligature case wasn't needed?

Reply:
@skef 
- Because currently, in feature files usage:
`sub f'lookup lig1 t' i';` has the same effect as `sub f'lookup lig1 t'lookup lig1 i'lookup lig1;`, where lookup lig1 has a rule `sub f t i by f_t_i`.
- Keeping the ligature case (the lines I commented out) makes FontForge prefer the latter approach. However, most .fea files I've seen use the former approach.